bazel/foreign_cc/0002-local-rate-limit-bucket-backport.patch

```diff
@@ -0,0 +1,3422 @@
+   bool local_rate_limit_per_downstream_connection = 11;
+ 
++  // Enables the local cluster level rate limiting, iff this is set explicitly. For example,
++  // given an Envoy gateway that contains N Envoy instances and a rate limit rule X tokens
++  // per second. If this is set, the total rate limit of whole gateway will always be X tokens
++  // per second regardless of how N changes. If this is not set, the total rate limit of whole
++  // gateway will be N * X tokens per second.
++  //
++  // .. note::
++  //   This should never be set if the ``local_rate_limit_per_downstream_connection`` is set to
++  //   true. Because if per connection rate limiting is enabled, we assume that the token buckets
++  //   should never be shared across Envoy instances.
++  //
++  // .. note::
++  //   This only works when the :ref:`local cluster name
++  //   <envoy_v3_api_field_config.bootstrap.v3.ClusterManager.local_cluster_name>` is set and
++  //   the related cluster is defined in the bootstrap configuration.
++  common.ratelimit.v3.LocalClusterRateLimit local_cluster_rate_limit = 16;
++
+   // Defines the standard version to use for X-RateLimit headers emitted by the filter.
+   //
+   // Disabled by default.
+diff --git a/envoy/ratelimit/ratelimit.h b/envoy/ratelimit/ratelimit.h
+index 2839cfa86c..5854372136 100644
+--- envoy/ratelimit/ratelimit.h
++++ envoy/ratelimit/ratelimit.h
+@@ -49,23 +49,38 @@ struct Descriptor {
+   absl::optional<RateLimitOverride> limit_ = absl::nullopt;
+ };
+ 
+-/**
+- * A single token bucket. See token_bucket.proto.
+- */
+-struct TokenBucket {
+-  uint32_t max_tokens_;
+-  uint32_t tokens_per_fill_;
+-  absl::Duration fill_interval_;
+-};
+-
+ /**
+  * A single rate limit request descriptor. See ratelimit.proto.
+  */
+ struct LocalDescriptor {
+   std::vector<DescriptorEntry> entries_;
+-  friend bool operator==(const LocalDescriptor& lhs, const LocalDescriptor& rhs) {
+-    return lhs.entries_ == rhs.entries_;
++
++  friend bool operator==(const LocalDescriptor& a, const LocalDescriptor& b) {
++    return a.entries_ == b.entries_;
+   }
++  struct Hash {
++    using is_transparent = void; // NOLINT(readability-identifier-naming)
++    size_t operator()(const LocalDescriptor& d) const {
++      return absl::Hash<std::vector<DescriptorEntry>>()(d.entries_);
++    }
++  };
++  struct Equal {
++    using is_transparent = void; // NOLINT(readability-identifier-naming)
++    size_t operator()(const LocalDescriptor& a, const LocalDescriptor& b) const {
++      return a.entries_ == b.entries_;
++    }
++  };
++
++  std::string toString() const {
++    return absl::StrJoin(entries_, ", ", [](std::string* out, const auto& e) {
++      absl::StrAppend(out, e.key_, "=", e.value_);
++    });
++  }
++
++  /**
++   * Local descriptor map.
++   */
++  template <class V> using Map = absl::flat_hash_map<LocalDescriptor, V, Hash, Equal>;
+ };
+ 
+ /*
+diff --git a/lbucket.iff b/lbucket.iff
```

Comment:
i'm a bit unclear on what this file is doing in the pull request and what its role is. was this perhaps added by mistake?

Reply:
file was committed by accident